### PR TITLE
Prevent kernel_normalizer to change mask dtype

### PR DESCRIPTION
**Pull request by ddonatien**
## Motivation

When training fpn_carafe neck in mmdet with feature of dtype torch.half, the execution fails with:
_RuntimeError: expected scalar type Half but found Float_
at mmcv/ops/carafe.py l130

The code works with features of dtype torch.float

## Modification

The issue arise from  mismatch in dtype between features and masks. This difference is created in the 'kernel_normalizer' function (carafe.py l263) where F.softmax returns by default results of dtype torch.float.

By setting the dtype argument to mask.dtype, F.softmax forwards the correct dtype.

## BC-breaking (Optional)

No

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Thanks for your kind contribution. Would you like to leave comments there to indicate that it is for FP16 compatibility?

**Comment by ddonatien** [comment]
I'm not sure about fp16 compatibility, using a different backbone from the ones used in the carafe config folder in mmdet created the issue.

**Comment by OpenMMLab-Coodinator** [comment]
Hi @ddonatien ！First of all, we want to express our gratitude for your significant PR in this project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you are Chinese or have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
